### PR TITLE
Function implementation

### PR DESCRIPTION
**Pull request by rongrong**
Ready for pre-review on high level. I'll probably use some data structures introduced in #13484 and add tests once it's merged.

### CODE REVIEW

**Comment by rongrong** [comment]
@wenleix There are errors from `GrowthHashFunctions`. Can you take a look?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 1]
```diff
@@ -168,12 +169,7 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
      *
      * @throws PrestoException if there are no matches or multiple matches
      */
-    public FunctionHandle resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
```
Comment (caithagoras) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: What is preventing us from being able to pass in a Session object?

File: presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java [hunk 2]
```diff
@@ -120,17 +134,32 @@ public final FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
...
                 SCALAR,
+                function.getLanguage(),
                 function.isDeterministic(),
                 function.isCalledOnNullInput());
     }
 
+    protected static FunctionImplementation sqlInvokedFunctionToImplementation(SqlInvokedRegularFunction function)
+    {
+        checkArgument(function.getLanguage().equals(SQL));
+        return new SqlInvokedRegularSqlFunctionImplementation(function.getBody());
```
Comment (caithagoras) [inline presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java]: I don't see this class anywhere in the PR.

**Comment by caithagoras** [comment]
Currently, `FunctionLanguage` is used to distinguish between built-in and user-defined sql. Rather than tying in with `Language`, we should give it a different name since it fullfill a different purpose. Suppose we support `Python` and `JS` in the future, they'll be have different `Language`, but might have the same execution strategy of using an external server.

Some naming suggestion:
- `FunctionImplementationType` (most preferred)
- `FunctionImplementation.Type`
- `FunctionExecutionStrategy`

**Comment by caithagoras** [comment]
A few issue with the `FunctionImplementation` Interface.
1. Adding `FunctionManager.getFunctionImplementation` and `BuiltInFunctionNamespaceManager.getFunctionImplementation` seems to aim for unifying function implementation fetching as the names are very generic, but that also makes it confusing since they're only fetching scalar functions.
2. I checked all callers of `FunctionImplementation.getFunctionLanguage`. This method is never actually used to make decisions - it is only used to perform sanity checks, which should be achieved by properly designing of the interfaces.
3. `FunctionImplementation#getImplementation` returns an Object, and you always down-cast when using it. This indicates something can be improved (returning an empty interface is better than returning an Object even if there is still a down-cast).

Here is my proposal:
a. All decisions are made using `FunctionMetadata.getFunctionImplemenationType` if needed (you're already doing this)
b(1) Either introduce a new method `FunctionNamespace.getSqlInvokedFunctionImplementation` does all the checks and thus can return a `SqlInvokedRegularFunctionImplemenation`.
Or
b(2). `FunctionNamespace(Manager).getFunctionImplementation` needs to consider scalar, window, aggregation, to at least match the expected behavior as indicated by the method name.
c. Remove the 2 methods (due to 2 and 3) in `FunctionImplementation` to make it an empty interface.
  - When you have `checkArgument(implementation.getFunctionLanguage() == SQL)`, that means what you should really pass in is a `SqlInvokedRegularFunctionImplemenation` instance instead. (Most of them are in `SqlFunctionUtils`)
  - If you change all those parameters, I think you'll end up having to do one down-cast right after you fetch the implementation from `FunctionManager` (or no cast if you do b(1)), which I think is a better practice (to fail early in case the thing being fetched misaligns with the expectation).

**Comment by rongrong** [comment]
> In this PR, `FunctionLanguage` is used to distinguish between built-in and user-defined functions. Rather than tying it with `Language`, we should give it a different name since it fullfill a different purpose. For example, suppose we support `Python` and `JS` in the future, they'll have a different `Language`, but might still have the same execution strategy of using an external server.
> 
> Some naming suggestions:
> 
> * `FunctionImplementationType` (this sounds best to me)
> * `FunctionImplementation.Type`
> * `FunctionExecutionStrategy`

Ya, it makes less and less sense to have this `FunctionLanguage` especially after the syntax support. The "language" here should theoretically be associated with the "execution engine" (how to run them) so I'll take `FunctionExecutionStrategy`. If you have more ideas let me know. I'm terrible with names. 🤣

**Comment by rongrong** [comment]
> A few issues with the `FunctionImplementation` Interface.
> 
> * Adding `FunctionManager.getFunctionImplementation` and `BuiltInFunctionNamespaceManager.getFunctionImplementation` seems to aim for unifying function implementation fetching, since the names are very generic, but that also makes them confusing since they're only fetching scalar functions.
> * I checked all callers of `FunctionImplementation.getFunctionLanguage`. This method is never actually used to make decisions - it is only used to perform sanity checks, which can be achieved by properly designing the interfaces.
> * `FunctionImplementation#getImplementation` returns an Object. This indicates something can be improved (even returning an empty interface is better than that).
> 
> Here is my proposal:
> 
> * All decisions are made using `FunctionMetadata.getFunctionImplemenationType` if needed (you're already doing this in this PR)
> * Either
>   (*) Introduce a new method `FunctionNamespace.getSqlInvokedFunctionImplementation`, mimicking the 3 existing getters, doing the checks and the down cast, and returning a `SqlInvokedRegularFunctionImplemenation`.
>   Or
>   (**) `FunctionManager.getFunctionImplementation` and `FunctionNamespaceManager.getFunctionImplementation` needs to include scalar, aggregation, and window, to at least match the expected behavior as indicated by the method name.
> * Remove the 2 methods in `FunctionImplementation` to make it an empty
>   interface.
>   
>   * When you have `checkArgument(implementation.getFunctionLanguage() == SQL)`, that means what you should really pass in instead is a `SqlInvokedRegularFunctionImplemenation` instance. (Most of cases are in `SqlFunctionUtils`)
>   * If you change all those parameters, I think you'll end up having to do one down-cast right after you fetch the implementation from `FunctionManager` if you do (**), or no cast at all if you do (*). And I think it is a better practice to cast early in order to fail early.

We sort of had a conversation about unifying the metadata and implementation but seems like they should still be two separate interfaces. At the point I'm inclined to have the implementation to be an empty interface. I'm thinking of maybe calling it `ScalarFunctionImplementation` to get away with the integration / window function issues. Problem with unifying them all is that their internal representation for builtin functions are very different.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 3]
```diff
@@ -893,10 +893,12 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
             }
             FunctionHandle functionHandle = metadata.getFunctionManager().resolveFunction(session, node.getName(), fromTypes(argumentTypes));
             FunctionMetadata functionMetadata = metadata.getFunctionManager().getFunctionMetadata(functionHandle);
-            for (int i = 0; i < argumentValues.size(); i++) {
-                Object value = argumentValues.get(i);
-                if (value == null && !functionMetadata.isCalledOnNullInput()) {
-                    return null;
+            if (!functionMetadata.isCalledOnNullInput()) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Might need to change `RowExpressionInterpreter::visitCall` as well

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Code in `RowExpressionInterpreter::visitCall` was structured a little differently so it's less weird. Anyways I added another PR to change that. You can see whether you want to keep that or not.

**Comment by rongrong** [comment]
Addressed comments, please review.

File: presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/SqlInvokedRegularFunction.java [hunk 4]
```diff
@@ -125,6 +136,12 @@ public RoutineCharacteristics getRoutineCharacteristics()
         return routineCharacteristics;
     }
 
+    public FunctionImplementationType getFunctionImplementationType()
+    {
+        checkState(LANGUAGE_TO_IMPLEMENTATION_MAP.containsKey(routineCharacteristics.getLanguage()), "Does not support language %s", routineCharacteristics.getLanguage());
```
Comment (arhimondr) [inline presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/SqlInvokedRegularFunction.java]: nit: `language is not supported: %s` ?

File: presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionProperties.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+        public Builder setLegacyRowFieldOrdinalAccessEnabled(boolean legacyRowFieldOrdinalAccessEnabled)
+        {
+            this.legacyRowFieldOrdinalAccessEnabled = legacyRowFieldOrdinalAccessEnabled;
+            return this;
+        }
+
+        public Builder setTimeZoneKey(TimeZoneKey timeZoneKey)
+        {
+            this.timeZoneKey = timeZoneKey;
```
Comment (arhimondr) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionProperties.java]: nit: `requireNonNull`

File: presto-main/src/main/java/com/facebook/presto/sql/gen/RowExpressionCompiler.java [hunk 6]
```diff
@@ -86,14 +100,36 @@ public BytecodeNode compile(RowExpression rowExpression, Scope scope, Optional<V
...
+                    // generate bytecode for SQL function
+                    return new RowExpressionCompiler(classDefinition, callSiteBinder, cachedInstanceBinder, fieldReferenceCompiler, metadata, sqlFunctionProperties, newCompiledLambdaMap.build()).compile(
+                            function,
+                            context.getScope(),
+                            context.getOutputBlockVariable(),
+                            context.getLambdaInterface());
 
-            return (new FunctionCallCodeGenerator()).generateCall(call.getFunctionHandle(), generatorContext, call.getType(), call.getArguments(), context.getOutputBlockVariable());
+                default:
+                    throw new UnsupportedOperationException();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/RowExpressionCompiler.java]: nit: `throw new IllegalArgumentException("Unsupported implementation type" ...`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 7]
```diff
@@ -906,7 +910,27 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
...
+                    ExpressionInterpreter functionInterpreter = new ExpressionInterpreter(
+                            function,
+                            metadata,
+                            session,
+                            getExpressionTypes(session, metadata, new SqlParser(), TypeProvider.empty(), function, emptyList(), WarningCollector.NOOP),
+                            optimize);
+                    result = functionInterpreter.visitor.process(function, context);
+                    break;
+                default:
+                    throw new PrestoException(NOT_SUPPORTED, "not supported");
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java [hunk 8]
```diff
@@ -259,7 +262,25 @@ public Object visitCall(CallExpression node, Object context)
...
+                    RowExpressionInterpreter rowExpressionInterpreter = new RowExpressionInterpreter(function, metadata, session, optimizationLevel);
+                    if (optimizationLevel.ordinal() >= EVALUATED.ordinal()) {
+                        value = rowExpressionInterpreter.evaluate();
+                    }
+                    else {
+                        value = rowExpressionInterpreter.optimize();
+                    }
+                    break;
+                default:
+                    throw new UnsupportedOperationException(format("Function implementation type %s is not supported", functionMetadata.getImplementationType()));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 9]
```diff
@@ -189,8 +185,8 @@ private FunctionHandle resolveFunction(Optional<Session> session, QualifiedName
             throw new PrestoException(FUNCTION_NOT_FOUND, format("Cannot find function namespace for function %s", name));
         }
 
-        Optional<FunctionNamespaceTransactionHandle> transactionHandle = session.flatMap(Session::getTransactionId)
-                .map(transactionId -> transactionManager.getFunctionNamespaceTransaction(transactionId, functionNamespaceManager.get().getName()));
+        Optional<FunctionNamespaceTransactionHandle> transactionHandle = transactionId
+                .map(tid -> transactionManager.getFunctionNamespaceTransaction(tid, functionNamespaceManager.get().getName()));
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java]: nit: `tid` -> `id`

File: presto-spi/src/main/java/com/facebook/presto/spi/function/ScalarFunctionImplementation.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+public interface ScalarFunctionImplementation
+{
+    /**
+     * We use FunctionImplementationType from FunctionMetadata to know the actual type of implementation object.
+     * The FunctionImplementationType -> Object type mapping is:
+     * BUILTIN -> BuiltInScalarFunctionImplementation
+     * SQL     -> String
+     */
+
+    Object getImplementation();
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/ScalarFunctionImplementation.java]: This design pattern doesn't look at an "Interface design pattern" but more like a "delegation". `FunctionImplementationType` should be good enough for caller to decide who to delegate. This interface may not be necessary. In the worst case, presto-main depends on presto-sql-function, `parseSqlFunctionExpression` can down cast it:
```java
((SqlInvokedRegularSqlFunctionImplementation) functionImplementation).getImplementation()
```
where only `SqlInvokedRegularSqlFunctionImplementation` needs a `String getImplementation()` method.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        }
+        return typeBuilder.build();
+    }
+
+    private static Map<Identifier, VariableReferenceExpression> buildIdentifierToVariableMap(FunctionMetadata functionMetadata, Expression sqlFunction, SqlFunctionProperties sqlFunctionProperties, Metadata metadata, PlanVariableAllocator variableAllocator)
+    {
+        // Allocate variables for identifiers
+        Map<String, Type> argumentTypes = getFunctionArgumentTypes(functionMetadata, metadata);
+        Map<NodeRef<Expression>, Type> expressionTypes = getSqlFunctionExpressionTypes(metadata, sqlFunctionProperties, sqlFunction, argumentTypes);
+        Map<Identifier, VariableReferenceExpression> variables = new LinkedHashMap();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/SqlFunctionUtils.java]: new LinkedHashMap<>();

**Comment by rongrong** [comment]
Addressed comments. Please review. Thanks! @caithagoras @highker

**Comment by caithagoras** [comment]
Commit title `Rename ScalarFunctionImplementation to BuiltInScalarFunctionImplement…` exceeds 72 characters and has been cut off.

**Comment by rongrong** [comment]
> Commit title `Rename ScalarFunctionImplementation to BuiltInScalarFunctionImplement…` exceeds 72 characters and has been cut off.

What can I change it to? 🤣

**Pull request merged**